www/js/lib/zimArchive.js

```diff
@@ -142,31 +142,68 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
      */
 
     /**
-     * Look for DirEntries with title starting with the given prefix.
+     * Look for DirEntries with title starting with the prefix of the current search object.
      * For now, ZIM titles are case sensitive.
      * So, as workaround, we try several variants of the prefix to find more results.
      * This should be enhanced when the ZIM format will be modified to store normalized titles
      * See https://phabricator.wikimedia.org/T108536
      * 
-     * @param {String} prefix
-     * @param {Integer} resultSize
-     * @param {callbackDirEntryList} callback
+     * @param {Object} search The current state.searches object
+     * @param {Integer} resultSize The number of dirEntries to find
+     * @param {callbackDirEntryList} callback The function to call with the result
+     * @param {Boolean} noInterim A flag to prevent callback until all results are ready (used in testing) 
      */
-    ZIMArchive.prototype.findDirEntriesWithPrefix = function(prefix, resultSize, callback) {
+    ZIMArchive.prototype.findDirEntriesWithPrefix = function (search, resultSize, callback, noInterim) {
+        var prefix = search.prefix;
         var that = this;
-        var prefixVariants = util.removeDuplicateStringsInSmallArray([prefix, util.ucFirstLetter(prefix), util.lcFirstLetter(prefix), util.ucEveryFirstLetter(prefix)]);
+        // Establish array of initial values that must be searched first
+        var startArray = [];
+        // Ensure a search is done on the string exactly as typed
+        startArray.push(prefix);
+        // Normalize any spacing and make string all lowercase
+        prefix = prefix.replace(/\s+/g, ' ').toLocaleLowerCase();
+        // Add lowercase string with initial uppercase (this is a very common pattern)
+        startArray.push(prefix.replace(/^./, function (m) {
+            return m.toLocaleUpperCase();
+        }));
+        // Get the full array of combinations to check number of combinations
+        var fullCombos = util.removeDuplicateStringsInSmallArray(util.allCaseFirstLetters(prefix, 'full'));
+        // Put cap on exponential number of combinations (five words = 3^5 = 243 combinations)
+        search.type = fullCombos.length < 200 ? 'full' : 'basic';
+        // We have to remove duplicate string combinations because util.allCaseFirstLetters() can return some combinations
+        // where uppercase and lowercase combinations are exactly the same, e.g. where prefix begins with punctuation
+        // or currency signs, for languages without case, or where user-entered case duplicates calculated case
+        var prefixVariants = util.removeDuplicateStringsInSmallArray(
+            startArray.concat(
+                // Get basic combinations first for speed of returning results
+                util.allCaseFirstLetters(prefix).concat(
+                    search.type === 'full' ? fullCombos : []
```

Comment:
When the search type is `full` (which will be the most frequent scenario), it calls `util.allCaseFirstLetters(prefix)` twice (with and without the `full` parameter), concatenates and de-duplicates the result.
Wouldn't be possible to use only the result of `util.allCaseFirstLetters(prefix,'full')`, and modify this `allCaseFirstLetters` function, so that the basic combinations are at the beginning of its result?
It would probably save some CPU

Reply:
I considered this, but it seemed like quite a lot of coding to get the order right when just running it twice does this for us. Basic search combinations are much more probable than full combinations, hence results are returned much faster to the UI if we search basic followed by any full combinations that were not already in basic. Because of the way the loop runs in the full type, it is impossible to guarantee that all the basic ones are returned first, whether you count up from 0000 to 1111 or whether you count down from 1111 to 0000.

And CPU usage here is tiny compared to binary search itself, so this is never going to be the bottleneck...

Reply:
All right, I trust you